### PR TITLE
feat: add on_checkpoint_end hook to pytorch trial

### PR DESCRIPTION
**Pull request by sidneyw**
# Test Plan
- UI change:
  - [ ] add screenshots
  - [ ] React? build & check storybooks
- [ ] user-facing api change: modify documentation and examples
- [ ] user-facing api change: add the "User-facing API Change" label
- [ ] bug fix: add regression test
- [ ] bug fix: determine if there are other similar bugs in the codebase
- [ ] new feature: add test coverage for any user-facing aspects
- [ ] refactor: maintain existing code coverage

### CODE REVIEW

File: harness/determined/pytorch/_pytorch_trial.py [hunk 1]
```diff
@@ -795,6 +796,13 @@ def create_lr_scheduler(
         """
         pass
 
+    def on_checkpoint_end(self, path: pathlib.Path) -> None:
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:799]: maybe: rename `path` to `checkpoint_dir`

**Comment by sidneyw** [comment]
Closing in favor of #405
